docs/pallas/tpu/collectives.rst

```diff
@@ -0,0 +1,452 @@
+Pallas Collectives
+===================
+Pallas has support for manual DMA (between memories of the same chip) and remote DMA (between memories of different chips) support. It also has an API for nesting pipelines of compute overlapped collectives.
+
+Manual DMAs are a necessary imperative escape hatch for maximum performance, but we hope that nested pipelines will minimize where they are used in codebases.
+
+Manual DMAs
+------------
+
+.. code-block:: python
+
+  from jax.experimental import pallas as pl
+  from jax.experimental.pallas import tpu as pltpu
+
+  def move_data(src_ref, dst_ref, sem, send_sem, recv_sem):
+    local_copy = pltpu.async_copy(src_ref, dst_ref.at[0:8], sem)
+    local_copy.wait()
+
+    # Should do a barrier the first time we communicate
+    # with another chip.
+    barrier_sem = pltpu.get_barrier_semaphore()
+    pltpu.semaphore_signal(barrier_sem, device_id=neighbor)
+    pltpu.semaphore_wait(barrier_sem)
+
+    dev_id = pltpu.device_id()
+    other_dev_id = 1 - dev_id
+    remote_copy = pltpu.async_remote_copy(
+      src_ref, dst_ref.at[0:8], send_sem, recv_sem, other_dev_id,
+      device_id_type=pltpu.DeviceIdType.LOGICAL,
+    )
+    # remote_copy.wait() for both.
+    remote_copy.wait_send()
+    remote_copy.wait_recv()
+
+    # Both APIs have make_ variants for building a descriptor without starting
+    # it. This allows us to start a DMA in one grid iteration and wait for it
+    # in another.
+    local_copy = pltpu.make_async_copy(src_ref, dst_ref.at[0:8], sem)
+    local_copy.start()
+    local_copy.wait()
+  
+  src_data = pltpu.VMEM((8,128), jnp.float32)
+  dst_data = pltpu.VMEM((16,128), jnp.float32)
+  semaphore = pltpu.SemaphoreType.DMA
+
+  # Inside a kernel:
+  pltpu.run_scoped(
+    move_data, src_data, dst_data, *((semaphore,) * 3)
+  )
+
+  # Or alternatively to use data that is live across iterations:
+  pl.pallas_call(
+      move_data,
+      grid_spec=pltpu.PrefetchScalarGridSpec(
+          num_scalar_prefetch=0,
+          # Could also use other memory spaces and input/outputs.
+          scratch_shapes=[src_data, dst_data, semaphore],
+      ),
+  )()
+
+.. warning::
+   Care must be taken when using DMAs to write to inputs or outputs as the reads and writes to these are pipelined with the execution of the body of your kernel. The pipeline double buffering will cause undefined behavior if you try to carry out a DMA across grid iterations where the block windows/indices of the operands are changing. Modifying inputs is also generally not supported as there is no mechanism for the writes to make it to the source buffer from the current ref of the current block.
```

Comment:
What, exactly is the contract?

Perhaps something like:
"Manual DMAs to the inputs or outputs of a pipeline is undefined behavior: there are no guarantees about the ordering of manual DMAs with respect to the pipeline's reads and writes."